src/ItemGrid.cpp

```diff
@@ -791,3 +863,56 @@ void cItemGrid::TriggerListeners(int a_SlotNum)
 
 
 
+
+bool cItemGrid::IsCompletelyEmpty() const
+{
+	return (m_Slots == nullptr);
+}
+
+
+
+
+
+const cItem & cItemGrid::UncheckedGetSlot(int a_SlotNum) const
+{
+	if (m_Slots == nullptr)
+	{
+		// An unallocated grid must be empty
+		static const cItem EmptyItem;
+		return EmptyItem;
+	}
+	else
+	{
+		return m_Slots[static_cast<size_t>(a_SlotNum)];
+	}
+}
+
+
+
+
+
+cItem * cItemGrid::GetSlots()
+{
+	if (m_Slots == nullptr)
+	{
+		m_Slots.reset(new cItem[m_NumSlots]);
+	}
+	return m_Slots.get();
+}
+
+
+
+
+
+const cItem * cItemGrid::GetSlots() const
```

Comment:
I would never return a `std::unique_ptr<T>&` here (maybe even ever) because there is no implied ownership.  Something like an `array_view` would probably be better but a raw pointer is still appropiate.

`GetSlot` would be a poor name considering there's already a function with that name with a different return type.  Anyway, the main reason for `GetSlots` is avoiding duplicated checks for `m_Slots` being `nullptr`.

Reply:
Just a side question: Why is there no implied ownership? Because we don't change ownership with the reference. And as long, as nobody has the great idea of saving the reference in a member variable 👿 , we should be fine.
And we get the benefits, that no broken code could ever think, he needs to destroy this pointer: `std::unique_ptr<> my_slots(GetSlots())`

Reply:
`GetSlots` was simply meant to observe the slots, just like `data()` currently. If instead you return `std::unique_ptr<T>&` then the caller could "steal" ownership:
```cpp
auto StolenSlots = std::move(Array.GetSlots());
```
I don't want the return value of `GetSlots` to imply that taking ownership is a valid option so I don't think `std::unique_ptr<T>&` is appropriate.

As a general philosophy, `T *` shouldn't be used to transfer ownership in modern code anyway.  In this slightly idealised frame of mind it's obvious that you shouldn't delete the pointer.  At least those are my thoughts.